File: Editorial/0000-0100/0014_Longest_Common_Prefix/C/3_Divide_and_conquer.c
char* LCP(char* s1, char* s2) {
    int minLen = strlen(s1) < strlen(s2) ? strlen(s1) : strlen(s2);
    char* res = (char*)calloc(minLen + 1, sizeof(char));
    for (int i = 0; i < minLen; i++) {
        if (s1[i] != s2[i]) {
            res[i] = '\0';
            break;
        }
        res[i] = s1[i];
    }
    return res;
}

char* divide_and_conquer(char** strs, int l, int r) {
    if (l == r)
        return strs[l];
    else {
        int mid = (l + r) / 2;
        char* lcpLeft = divide_and_conquer(strs, l, mid);
        char* lcpRight = divide_and_conquer(strs, mid + 1, r);
        char* res = LCP(lcpLeft, lcpRight);
        if (lcpLeft != strs[l]) free(lcpLeft);
        if (lcpRight != strs[mid + 1]) free(lcpRight);
        return res;
    }
}

char* longestCommonPrefix(char** strs, int strsSize) {
    if (strs == NULL || strsSize == 0) return "";
    return divide_and_conquer(strs, 0, strsSize - 1);
}

File: Editorial/0000-0100/0014_Longest_Common_Prefix/Go/3_Divide_and_conquer.go
func longestCommonPrefix(strs []string) string {
    if strs == nil || len(strs) < 1 {
        return ""
    }

    var LCP func(string, string) string
    LCP = func(str1, str2 string) string {
        minLen := min(len(str1), len(str2))
        for i := 0; i < minLen; i++ {
            if str1[i] != str2[i] {
                return str1[0:i]
            }
        }
        return str1[0:minLen]
    }

    var divideAndConquer func([]string, int, int) string
    divideAndConquer = func(strs []string, l, r int) string {
        if l == r {
            return strs[l]
        } else {
            mid := (l + r) / 2
            lcpLeft := divideAndConquer(strs, l, mid)
            lcpRight := divideAndConquer(strs, mid+1, r)
            return LCP(lcpLeft, lcpRight)
        }
    }

    return divideAndConquer(strs, 0, len(strs)-1)
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
